File: cpp/232_Implement_Queue_using_Stacks.py
class MyQueue:

    def __init__(self):
        self.ins = []
        self.out = []

    def move(self):
        if not self.out:
            while self.ins:
                self.out.append(self.ins.pop())

    def push(self, x: int) -> None:
        self.ins.append(x)

    def pop(self) -> int:
        self.move()
        return self.out.pop()

    def peek(self) -> int:
        self.move()
        return self.out[-1]

    def empty(self) -> bool:
        return not self.ins and not self.out

File: cpp/232_Implement_Queue_using_Stacks.cpp
#include <stack>
using namespace std;

class MyQueue {
private:
    stack<int> in;
    stack<int> out;

    void move() {
        if (out.empty()) {
            while (!in.empty()) {
                out.push(in.top());
                in.pop();
            }
        }
    }

public:
    MyQueue() {}

    void push(int x) {
        in.push(x);
    }

    int pop() {
        move();

        int val = out.top();
        out.pop();

        return val;
    }

    int peek() {
        move();
        return out.top();
    }

    bool empty() {
        return in.empty() && out.empty();
    }
};
